### PR TITLE
poetry: fix build and update from 1.0.5 -> 1.0.9

### PR DESCRIPTION
**Pull request by mweinelt**
###### Motivation for this change

Poetry failed to build.

- Downgrade `keyring`, could also be substituted by using a patch, since it's a multiline statement in pyproject.tml, if that is what you prefer.
  ```
   keyring = [
       { version = "^18.0.1", python = "~2.7 || ~3.4" },
       { version = "^20.0.1", python = "^3.5" }
   ]
   ```
- The clikit dependency now requires crashtest.
- I tried building crashtest from GitHub in pyproject format, but that failed somewhere in poetry, so it's falling back to an imports check now.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

@jbaum98

### CODE REVIEW

File: pkgs/development/python-modules/poetry/default.nix [hunk 1]
```diff
@@ -30,22 +30,34 @@
...
 
+  keyring' = keyring.overridePythonAttrs (old: rec {
+    version = "20.0.1";
+    src= fetchPypi {
+      inherit (old) pname;
+      inherit version;
+      sha256 = "138dbl7931i3gxm1fvvw5b7x5nnrbyg5h9ayvh5x6s8215zzlfwn";
+    };
+  });
+
```
Comment (jonringer) [inline pkgs/development/python-modules/poetry/default.nix:41]: this is highly discouraged please try to remove

File: pkgs/development/python-modules/crashtest/default.nix [hunk 2]
```diff
@@ -0,0 +1,24 @@
+{ lib, buildPythonPackage, fetchFromGitHub, fetchPypi, pytest }:
```
Comment (jonringer) [inline pkgs/development/python-modules/crashtest/default.nix:1]:
```suggestion
{ lib, buildPythonPackage, fetchFromGitHub, fetchPypi, isPy27, pytest }:
```

File: pkgs/development/python-modules/crashtest/default.nix [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+
+buildPythonPackage rec {
+  pname = "crashtest";
+  version = "0.3.0";
```
Comment (jonringer) [inline pkgs/development/python-modules/crashtest/default.nix:5]:
```suggestion
  version = "0.3.0";
  disabled = isPy27;
```
```
builder for '/nix/store/6c8w8sl3fm4cnlkyd07mk23pix203rrk-python2.7-crashtest-0.3.0.drv' failed with exit code 1; last 10 log lines:
  adding 'crashtest-0.3.0.dist-info/top_level.txt'
  adding 'crashtest-0.3.0.dist-info/RECORD'
  removing build/bdist.linux-x86_64/wheel
  Finished executing setuptoolsBuildPhase
  installing
  Executing pipInstallPhase
  /build/crashtest-0.3.0/dist /build/crashtest-0.3.0
  DEPRECATION: Python 2.7 reached the end of its life on January 1st, 2020. Please upgrade your Python as Python 2.7 is no longer maintained. pip 21.0 will drop support for Python 2.7 in January 2021. More details about Python 2 support in pip, can be found at https://pip.pypa.io/en/latest/development/release-process/#python-2-support
  Processing ./crashtest-0.3.0-py2-none-any.whl
  ERROR: Package 'crashtest' requires a different Python: 2.7.18 not in '>=3.6,<4.0'
```

**Review by jonringer — COMMENTED** [review]
both crashtest and keyring have removed support for python2, I would just remove it from poetry altogether. Also, we are quite outside the version range of ~1/2 the dependencies, we should have tests. This is what I came up with:

```
{ lib, buildPythonPackage, fetchFromGitHub, isPy27, pythonOlder
, cachecontrol
, cachy
, cleo
, clikit
, html5lib
, httpretty
, importlib-metadata
, intreehooks
, jsonschema
, keyring
, lockfile
, pexpect
, pkginfo
, pygments
, pyparsing
, pyrsistent
, pytestCheckHook
, pytestcov
, pytest-mock
, requests
, requests-toolbelt
, shellingham
, tomlkit
}:

buildPythonPackage rec {
  pname = "poetry";
  version = "1.0.9";
  format = "pyproject";
  disabled = isPy27;

  src = fetchFromGitHub {
    owner = "python-poetry";
    repo = pname;
    rev = version;
    sha256 = "0gi1li55rim60hf1gdpgpx84zlkaj0wv12wbv7dib9malhfj3pnz";
  };

  postPatch = ''
    substituteInPlace pyproject.toml \
     --replace "pyrsistent = \"^0.14.2\"" "pyrsistent = \"^0.16.0\"" \
     --replace "requests-toolbelt = \"^0.8.0\"" "requests-toolbelt = \"^0.9.1\"" \
     --replace 'importlib-metadata = {version = "~1.1.3", python = "<3.8"}' \
       'importlib-metadata = {version = ">=1.3,<2", python = "<3.8"}' \
     --replace "tomlkit = \"^0.5.11\"" "tomlkit = \"^0.6.0\"" \
     --replace "cleo = \"^0.7.6\"" "cleo = \"^0.8.0\"" \
     --replace "version = \"^20.0.1\", python = \"^3.5\"" "version = \"^21.0.0\", python = \"^3.5\"" \
     --replace "clikit = \"^0.4.2\"" "clikit = \"^0.6.2\""
  '';

  nativeBuildInputs = [ intreehooks ];

  propagatedBuildInputs = [
    cachecontrol
    cachy
    cleo
    clikit
    html5lib
    jsonschema
    keyring
    lockfile
    pexpect
    pkginfo
    pyparsing
    pyrsistent
    requests
    requests-toolbelt
    shellingham
    tomlkit
  ] ++ lib.optionals (pythonOlder "3.8") [ importlib-metadata ];

  postInstall = ''
    mkdir -p "$out/share/bash-completion/completions"
    "$out/bin/poetry" completions bash > "$out/share/bash-completion/completions/poetry"
    mkdir -p "$out/share/zsh/vendor-completions"
    "$out/bin/poetry" completions zsh > "$out/share/zsh/vendor-completions/_poetry"
    mkdir -p "$out/share/fish/vendor_completions.d"
    "$out/bin/poetry" completions fish > "$out/share/fish/vendor_completions.d/poetry.fish"
  '';

  checkInputs = [ pytestCheckHook httpretty pytest-mock pygments pytestcov ];
  preCheck = "export HOME=$TMPDIR";
  disabledTests = [
    # touches network
    "git"
    "solver"
    "load"
    "vcs"
    "prereleases_if_they_are_compatible"
    # requires git history to work correctly
    "default_with_excluded_data"
  ];

  meta = with lib; {
    homepage = "https://python-poetry.org/";
    description = "Python dependency management and packaging made easy";
    license = licenses.mit;
    maintainers = with maintainers; [ jakewaksbaum ];
  };
}
```

File: pkgs/development/python-modules/poetry/default.nix (hunk 1 above)
Comment (mweinelt) [reply pkgs/development/python-modules/poetry/default.nix:41]: Will add a patch and use substituteAll instead.

Comment (mweinelt) [reply pkgs/development/python-modules/poetry/default.nix:41]: nvm.

**Comment by mweinelt** [comment]
Yeah, thanks for the recommendation and your work on the tests. I squashed them into the poetry commit.

**Review by jonringer — APPROVED** [review]
LGTM

Result of nixpkgs-review pr 91222 [1](https://github.com/Mic92/nixpkgs-review)<details>  <summary>13 packages built:</summary><br>- cmake-language-server<br>- python37Packages.aria2p<br>- python37Packages.cleo<br>- python37Packages.clikit<br>- python37Packages.crashtest<br>- python37Packages.poetry<br>- python37Packages.xdg<br>- python38Packages.aria2p<br>- python38Packages.cleo<br>- python38Packages.clikit<br>- python38Packages.crashtest<br>- python38Packages.poetry<br>- python38Packages.xdg</details>

**Pull request merged**